src/backend/optimizer/path/allpaths.c

```diff
@@ -4434,6 +4447,141 @@ is_query_contain_limit_groupby(Query *parse)
 	return false;
 }
 
+typedef struct check_dml_context
+{
+	PlannerInfo *cteroot;
+	Index		m_query_level;	/* absolute query level */
+}			check_dml_context;
+
+/*
+ * Returns true if any subquery of given Query has any non-SELECT commandType.
+ */
+static bool
+query_contains_dml_walker(Node *node, check_dml_context * ctx)
+{
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) node;
+
+		/*
+		 * Check if the rte is CTE, which has been defined at a higher query
+		 * level than entry level of last subquery we called the walker with.
+		 * In this case we would be able to iterate through cteList of higher
+		 * level and analyze corresponding CTE.
+		 *
+		 * rte->ctelevelsup is equal to number of levels up from the level
+		 * where rte is referenced to the level where its CTE is defined.
+		 * Difference ctx->m_query_level - rte->ctelevelsup stands for
+		 * absolute query level, where the CTE is defined. In case if this
+		 * level is higher than absolute level of last subquery with
+		 * initialized PlannerInfo (last CTE and its root, which the walker
+		 * was last called with), we are able to derive the root. Based on its
+		 * cteList it is possible to investigate the CTE. Otherwise, this CTE
+		 * should have already been processed by query_tree_walker.
+		 *
+		 * For e.x. WITH cte1 AS (INSERT into...) SELECT * FROM (WITH cte2 as
+		 * (SELECT * from cte1) SELECT * FROM (WITH cte3 as (SELECT * FROM
+		 * cte2) SELECT * FROM cte3 a join cte3 b using (i)) foo) bar;
+		 *
+		 * Here cte3's subquery is an entry point for the walker with
+		 * cteroot->query_level = 3. cte3 references the cte2, whose
+		 * ctelevelsup is 2. It means that cte2 is defined at 2 levels up from
+		 * the m_query_level = 4. This level is higher than level where cte3
+		 * is defined, because cte3 is defined at level 3 and cte2 is at level
+		 * 2. In this case we can derive parent root, whose cteList contains
+		 * query of cte2.
+		 */
+		if (rte->rtekind == RTE_CTE &&
+			rte->ctelevelsup < ctx->m_query_level &&
+			(ctx->m_query_level - rte->ctelevelsup) <= ctx->cteroot->query_level)
```

Comment:
rte->ctelevelsup is a relative quantity. It represents the number of querly levels up from the point where this RTE is refenced to the level where its CTE is defined. This rte can be at arbitraty level. Let's consider the example:
```
explain (costs off)
with cte as (
    insert into t1 select i, i * 100 from generate_series(1,5) i
    returning *
), cte1 as (select * from t1) select * from (
with cte1 as (select * from cte)
select * from (with cte2 as (select * from cte1) select * from cte2 a join cte2 b using (i)) foo) bar;
```
cte_query_contains_dml is called on cte2's query first. ctx->cteroot->query_level=3. When walker comes across the rte corresponding to cte1, we have: rte->ctelevelsup=2. And if we go up from cteroot on 2 levels we will get to the level  1. It's the level of cte with INSERT. And there we will find the cte1 which is not even referenced anywhere, and will not detect the insert operation. It means that we need to find strict level of cte1. And without introduction of `m_query_level` it is impossible.

Second example dedicated to the problem that  [root structure is defined](https://github.com/arenadata/gpdb/blob/adb-7.1.0/src/backend/optimizer/path/allpaths.c#L2866-L2872) only for the query level where the original cte is referenced and higher. In case of nested CTEs we do not know any subroot.
```
explain (costs off)
with cte as (
    insert into t1 select i, i * 100 from generate_series(1,5) i
    returning *
),
cte2 as (
with cte3 as (select * from cte)
select (select i from cte3) from t1
)
select * from  cte2 a join cte2 b using (i);
```
In this case rte->ctelevelsup =1 for cte3, but `ctx->cteroot->query_level=1` as well. And it will require more complex logic to validate.

>Can we get rid of the check_dml_context struct and simplify this condition? 
I doubt.